nomad/csi_endpoint.go

```diff
@@ -1052,6 +1051,197 @@ func (v *CSIVolume) ListExternal(args *structs.CSIVolumeExternalListRequest, rep
 	return nil
 }
 
+func (v *CSIVolume) CreateSnapshot(args *structs.CSISnapshotCreateRequest, reply *structs.CSISnapshotCreateResponse) error {
+
+	if done, err := v.srv.forward("CSIVolume.CreateSnapshot", args, args, reply); done {
+		return err
+	}
+	defer metrics.MeasureSince([]string{"nomad", "volume", "create_snapshot"}, time.Now())
+
+	allowVolume := acl.NamespaceValidator(acl.NamespaceCapabilityCSIWriteVolume)
+	aclObj, err := v.srv.WriteACLObj(&args.WriteRequest, false)
+	if err != nil {
+		return err
+	}
+	if !allowVolume(aclObj, args.RequestNamespace()) || !aclObj.AllowPluginRead() {
+		return structs.ErrPermissionDenied
+	}
+
+	state, err := v.srv.fsm.State().Snapshot()
+	if err != nil {
+		return err
+	}
+
+	method := "ClientCSI.ControllerCreateSnapshot"
+	var mErr multierror.Error
+	for _, snap := range args.Snapshots {
```

Comment:
This is probably a little overly cautious, but if a user sent a malformed http request with a null value in the snapshots array I think we get a nil pointer dereference. Not sure how we usually handle this, may be worth validating in the http layer 
```
{"Snapshots":[null],"Region":"","Namespace":"","SecretID":""}
```

Reply:
Good call. We can't really rely on client code to be correct for sure.

Reply:
(same thing for DeleteSnapshot)

Reply:
Follow-up, I just double-checked that JSON decoding on this and it's not something we can rely on there unfortunately: https://play.golang.org/p/RYY6w5cQXYC

Reply:
👍 I was able to get a panic by swapping out Snapshots with `[]*api.CSISnapshot{nil}` here https://github.com/hashicorp/nomad/pull/10245/files#diff-ad34cf50edc09fdf1b4573ba6b5ed49a0f9c2810a3ecefccdba6cfe797a5dd78R139

Reply:
Fixed in 5d7ba9f